### PR TITLE
r/sns_topic_subscription - recreate on topic deletion + validations

### LINKED ISSUE #9645: SNS topic subscription is not recreated after SNS topic delete/recreate
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
Terraform v0.11.14
provider.aws v2.22.0

### Affected Resource(s)

* aws_sns_topic_subscription
* aws_sns_topic

### Terraform Configuration Files

One directory with an SNS declaration :

```hcl
provider "aws" {
  region = "eu-west-1"
}

resource "aws_sns_topic" "hello" {
  name = "sns-topic"
}
```

Another directory with the SQS declaration

```hcl
provider "aws" {
  region = "eu-west-1"
}

data "aws_sns_topic" "sns-topic" {
  name = "sns-topic"
}

resource "aws_sqs_queue" "stream_out_queue" {
  name                       = "sqs-queue"
  visibility_timeout_seconds = "360"
}

resource "aws_sqs_queue_policy" "sns_to_sqs_policy" {
  queue_url = "${aws_sqs_queue.stream_out_queue.id}"
  policy    = "${data.aws_iam_policy_document.sns_to_sqs_policy_document.json}"
}


resource "aws_sns_topic_subscription" "stream-out_sqs_target" {
  topic_arn              = "${data.aws_sns_topic.sns-topic.arn}"
  protocol               = "sqs"
  endpoint               = "${aws_sqs_queue.stream_out_queue.arn}"
  endpoint_auto_confirms = true
}


data "aws_iam_policy_document" "sns_to_sqs_policy_document" {
  statement {
    actions = ["sqs:SendMessage"]

    resources = ["${aws_sqs_queue.stream_out_queue.arn}"]

    principals {
      type        = "AWS"
      identifiers = ["*"]
    }

    condition {
      test     = "ArnEquals"
      variable = "aws:SourceArn"

      values = ["${data.aws_sns_topic.sns-topic.arn}"]
    }
  }
}
```
### Debug Output
[Debug Output](https://gist.github.com/Tirke/4456548aa795c8480cfbb0221351d14c)

### Expected Behavior

When applying the code containing the `aws_sns_topic_subscription` declaration it should always recreate it if it is NOT present.

### Actual Behavior

`aws_sns_topic_subscription` is NOT recreated and attached when it isn't present in the topic subscription BUT present in the overall list of existing subscriptions.

### Steps to Reproduce

1. `terraform init` and `terraform apply` the SNS topic code
2. `terraform init` and `terraform apply` the SQS code
3. `terraform destroy` the SNS topic code
4. `terraform apply` the SNS topic code
5. `terraform apply` the SQS code 

Result after step 5:
Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

But there is no subscription under the SNS topic, the `terraform apply` of the SQS code is not recreating the correct subscription.

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #9645

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSNSTopicSubscription_'
--- PASS: TestAccAWSSNSTopicSubscription_basic (77.90s)
--- PASS: TestAccAWSSNSTopicSubscription_filterPolicy (175.19s)
--- PASS: TestAccAWSSNSTopicSubscription_deliveryPolicy (134.98s)
--- PASS: TestAccAWSSNSTopicSubscription_rawMessageDelivery (130.24s)
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingEndpoint (104.22s)
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingSecuredEndpoint (166.53s)
--- PASS: TestAccAWSSNSTopicSubscription_disappears (156.98s)
--- PASS: TestAccAWSSNSTopicSubscription_disappears_topic (83.49s)
```

### CODE REVIEW

**Comment by DrFaust92** [comment]
just noticed https://github.com/terraform-providers/terraform-provider-aws/pull/10496 exists. 

let me know if i should close mine or rebase after the aforementioned PR is merged

**Comment by bflad** [comment]
#10496 has been merged, so this is ready for rebase

**Comment by DrFaust92** [comment]
rebased + reran tests

File: aws/resource_aws_sns_topic_subscription.go [hunk 1]
```diff
@@ -152,11 +158,11 @@ func resourceAwsSnsTopicSubscriptionUpdate(d *schema.ResourceData, meta interfac
 }
 
 func resourceAwsSnsTopicSubscriptionRead(d *schema.ResourceData, meta interface{}) error {
-	snsconn := meta.(*AWSClient).snsconn
+	conn := meta.(*AWSClient).snsconn
 
 	log.Printf("[DEBUG] Loading subscription %s", d.Id())
```
Comment (ewbankkit) [inline aws/resource_aws_sns_topic_subscription.go:164]: Call `ListSubscriptionsByTopic` first?
See https://github.com/terraform-providers/terraform-provider-aws/issues/9645#issuecomment-716659747.

**Comment by DrFaust92** [comment]
rerun tests:

```
--- PASS: TestAccAWSSNSTopicSubscription_disappears (111.31s)
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingEndpoint (121.52s)
--- PASS: TestAccAWSSNSTopicSubscription_deliveryPolicy (127.22s)
--- PASS: TestAccAWSSNSTopicSubscription_filterPolicy (162.21s)
--- PASS: TestAccAWSSNSTopicSubscription_rawMessageDelivery (171.12s)
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingSecuredEndpoint (201.53s)
```

File: aws/resource_aws_sns_topic_subscription.go (hunk 1 above)
Comment (DrFaust92) [reply aws/resource_aws_sns_topic_subscription.go:164]: Thank, Added.

**Comment by DrFaust92** [comment]
```
--- PASS: TestAccAWSSNSTopicSubscription_disappears_topic (37.01s)
--- PASS: TestAccAWSSNSTopicSubscription_basic (87.44s)
--- PASS: TestAccAWSSNSTopicSubscription_disappears (93.81s)
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingEndpoint (103.16s)
--- PASS: TestAccAWSSNSTopicSubscription_redrivePolicy (120.52s)
--- PASS: TestAccAWSSNSTopicSubscription_rawMessageDelivery (121.38s)
--- PASS: TestAccAWSSNSTopicSubscription_filterPolicy (124.27s)
--- PASS: TestAccAWSSNSTopicSubscription_deliveryPolicy (126.26s)
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingSecuredEndpoint (134.03s)
```

**Comment by YakDriver** [comment]
@DrFaust92 Thank you for this pull request! The `aws_sns_topic_subscription` recently underwent substantial modernization in #14923. We apologize for the inconvenience that will cause with this PR. However, we look forward to reviewing these improvements rebased on the modernization.

**Review by gdavison — APPROVED** [review]
LGTM 🚀 

--- PASS: TestAccAWSSNSTopicSubscription_email (25.45s)
--- PASS: TestAccAWSSNSTopicSubscription_disappears_topic (27.48s)
--- PASS: TestAccAWSSNSTopicSubscription_filterPolicy (65.17s)
--- PASS: TestAccAWSSNSTopicSubscription_rawMessageDelivery (67.47s)
--- PASS: TestAccAWSSNSTopicSubscription_redrivePolicy (67.72s)
--- PASS: TestAccAWSSNSTopicSubscription_deliveryPolicy (67.79s)
--- PASS: TestAccAWSSNSTopicSubscription_basic (71.10s)
--- PASS: TestAccAWSSNSTopicSubscription_disappears (84.14s)
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingEndpoint (87.73s)
--- PASS: TestAccAWSSNSTopicSubscription_autoConfirmingSecuredEndpoint (124.39s)
--- PASS: TestAccAWSSNSTopicSubscription_firehose (144.02s)

**Pull request merged**